### PR TITLE
build: upgrade to go1.16

### LINKED ISSUE #21301: Update to go 1.16
https://github.com/influxdata/influxdb/blob/master/go.mod#L3

### PR DESCRIPTION
**Pull request by danxmoran**
Closes #21301 

Apart from bumping the version in `go.mod`, I had to:
1. Bump the MacOS SDK & osxcross versions we use for cross-building
2. Fix some warnings that have become errors
3. Rearrange a piece of CI to deal with changes to `go mod tidy` and `go mod download`

### CODE REVIEW

File: scripts/ci/xcc-builds/build-osxcross-clang.sh [hunk 1]
```diff
@@ -32,7 +32,7 @@ docker run --rm -i -v ${OUT_DIR}:/out -w /tmp ${BUILD_IMAGE} bash <<EOF
   git clone https://github.com/tpoechtrager/osxcross.git /usr/local/osxcross && \
     cd /usr/local/osxcross && \
     git checkout ${OSXCROSS_VERSION} && \
-    curl -L -o ./tarballs/MacOSX10.11.sdk.tar.xz https://macos-sdks.s3.amazonaws.com/MacOSX10.11.sdk.tar.xz && \
+    curl -L -o ./tarballs/MacOSX10.12.sdk.tar.xz https://storage.googleapis.com/influxdata-team-flux/macos-sdks/MacOSX10.12.sdk.tar.xz && \
```
Comment (lesam) [inline scripts/ci/xcc-builds/build-osxcross-clang.sh:35]: Does this mean we support fewer Mac versions now? That should be in the release notes

File: scripts/ci/xcc.sh [hunk 2]
```diff
@@ -11,7 +11,7 @@ GOARCH=${GOARCH:-$(go env GOARCH)}
 case "${GOOS}_${GOARCH}" in
   linux_amd64)   CC=musl-gcc ;;
   linux_arm64)   CC=aarch64-unknown-linux-musl-gcc ;;
-  darwin_amd64)  CC=x86_64-apple-darwin15-clang ;;
+  darwin_amd64)  CC=x86_64-apple-darwin16-clang ;;
```
Comment (lesam) [inline scripts/ci/xcc.sh:14]: Same question

File: scripts/ci/xcc-builds/build-osxcross-clang.sh (hunk 1 above)
Comment (danxmoran) [reply scripts/ci/xcc-builds/build-osxcross-clang.sh:35]: Will add a note, good catch

Comment (danxmoran) [reply scripts/ci/xcc-builds/build-osxcross-clang.sh:35]: Done

**Review by docmerlin — APPROVED** [review]
WOOT!

**Pull request merged**